### DISCUSSION TITLE
Config Rendering with VLAN Ports in VLAN Interface

### DISCUSSION THREAD

**Discussion by v0tti** [netbox-community/netbox#12573] [Help Wanted!] [answered]
Hi all,
I am currently playing around with config rendering. We are using some Zyxel GS-Series switches.
In the config the ports a VLAN is assigned on are specified in the VLAN configuration like:
```
vlan 1000 
  name Test 
  fixed 2,9 
  forbidden 1,3-8,10 
  untagged 1-8,10 
```

I created a interface named `vlan 1000` of type virtual. Additionally I created all physical interfaces and assigned the corresponding VLANs to them.

Now, how am I supposed to create those lists of `fixed`, `forbidden` and `untagged` VLANs within the VLAN configuration context? The virtual VLAN interface in NetBox does not know to which VLAN object it is assigned to. Is there any "recommended" way to fix this?

**Comment by v0tti** [comment, marked as answer]
For now I found a pretty hacky solution by assigning the VLAN to the virtual interface as untagged:
```
fixed "{{ (vlan.untagged_vlan.interfaces_as_tagged.exclude(type="virtual").filter(device=device)|list + vlan.untagged_vlan.interfaces_as_untagged.exclude(type="virtual").filter(device=device)|list)|unique|join(",") }}"
forbidden "{{ device.interfaces.exclude(type="virtual")|list|reject("in", (vlan.untagged_vlan.interfaces_as_tagged.exclude(type="virtual").filter(device=device)|list + vlan.untagged_vlan.interfaces_as_untagged.exclude(type="virtual").filter(device=device)|list))|unique|join(",") }}"
untagged "{{ vlan.untagged_vlan.interfaces_as_untagged.exclude(type="virtual").filter(device=device)|list|join(",") }}"
```

Is this the best solution to my problem?

**Reply by candlerb** [reply]
This is pretty much the same as #12569.

The difference is the "forbidden" section. It's unclear what is meant here: I can't see how ports 3-8 can be forbidden, at the same time as ports 1-8 are untagged.

Perhaps 'forbidden' means interfaces which are shut down?

```
{%- for v in device.site.vlans.all() %}
{%- set ift = ((device.interfaces.filter(tagged_vlans=v)|list) + (device.interfaces.filter(mode="tagged-all")|list)) | unique | sort(attribute="name") %}
{%- set ifu = device.interfaces.filter(untagged_vlan=v) | sort(attribute="name") -%}
{%- set ifd = device.interfaces.filter(enabled=False) %}
{%- if ift or ifu %}
vlan {{v.vid}}
  name {{v.name}}
{%- if ift %}
  fixed {{ ift|join(",") }}
{%- endif %}
{%- if ifu %}
  untagged {{ ifu|join(",") }}
{%- endif %}
{%- if ifd %}
  forbidden {{ ifd|join(",") }}
{%- endif %}
{%- endif %}
{%- endfor %}
```

Or is "forbidden" just the opposite set of VLANs to "fixed"? In that case, unless you can get away with sending `forbidden 1-24` followed by `fixed a,b,c,d` you might have to write your own python code to work out the holes that have been punched in the blocks of interfaces.

EDIT: I think your solution is better than mine. I wasn't aware of `vlan.interfaces_as_(un)tagged`

**Reply by v0tti** [reply]
Yes, "forbidden" is like the opposite of "fixed". There is another mode called "normal" which I never quite understood and never needed. But like this it is working for me.

**Reply by candlerb** [reply]
How about this then?

```
{%- set dif = device.interfaces.exclude(type="virtual") %}
{%- for v in device.site.vlans.all() %}
{%- set ift = ((dif.filter(tagged_vlans=v)|list) + (dif.filter(mode="tagged-all")|list)) | unique | sort(attribute="name") %}
{%- set iff = dif.exclude(mode="tagged-all").exclude(tagged_vlans=v).exclude(untagged_vlan=v) | sort(attribute="name") %}
{%- set ifu = device.interfaces.filter(untagged_vlan=v) | sort(attribute="name") -%}
{%- if ift or ifu %}
vlan {{v.vid}}
  name {{v.name}}
{%- if ift %}
  fixed {{ ift|join(",") }}
{%- endif %}
{%- if iff %}
  forbidden {{ iff|join(",") }}
{%- endif %}
{%- if ifu %}
  untagged {{ ifu|join(",") }}
{%- endif %}
{%- endif %}
{%- endfor %}
```
